Enable slow compression with QuickCompress

tcmoore3 opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

This PR changes the `target_box` member of `UpdaterQuickCompress` from a box object to a `VectorVariantBox` to enable a time-varying target box, which enables slow compression of the box in HPMC.

## Motivation and context

Resolves #1641 

## How has this been tested?

Added a validation test that uses QuickCompress with a `BoxVariant` and ensures that the box compression occurs no faster than what the time-varying target box dictates.

## Change log

```
`hoomd.hpmc.update.QuickCompress` now accepts a `hoomd.variant.box.BoxVariant` object for `target_box`, enabling slow compression of simulation boxes in HPMC simulations.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

tcmoore3 committed:
Enable slow compression with QuickCompress

tcmoore3 committed:
Add slow compress test

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/quick-compress-moving-target-box

tcmoore3 committed:
Update documentation

tcmoore3 committed:
Formatting

tommy-waltmann commented on hoomd/hpmc/update.py:
```diff
@@ -894,9 +897,11 @@ def __init__(self,
         param_dict = ParameterDict(max_overlaps_per_particle=float,
                                    min_scale=float,
-                                   target_box=hoomd.Box,
+                                   target_box=hoomd.variant.box.BoxVariant,
```
If this attribute is settable, we need to coerce a regular box into a box variant whenever the attribute is set, not just in the constructor.

joaander replied:
Yes, use a converter method here.

tcmoore3 replied:
I think I implemented the preprocessing correctly, but I'm running into the same import issues that I was in the vector variant PR. I will need help fixing that.

joaander replied:
Yes, this is a circular import:
`hoomd/variant/box.py`:
```
from hoomd.data.typeconverter import box_preprocessing, variant_preprocessing
```
`hoomd/data/typeconvertor.py`:
```
from hoomd.variant.box import Constant as ConstantBox
```

The solution is to not use `from ... import ...`. I pushed a commit that fixes this.

tommy-waltmann commented on hoomd/hpmc/pytest/test_quick_compress.py:
```diff
@@ -13,20 +13,35 @@
 valid_constructor_args = [
     dict(trigger=hoomd.trigger.Periodic(10),
-         target_box=hoomd.Box.from_box([10, 10, 10])),
+         target_box=hoomd.variant.box.Constant(hoomd.Box.from_box([10, 10,
+                                                                   10]))),
```
Let's keep at least one case where the constructor arg is still a regular box since it should be able to accept both.

tommy-waltmann commented on hoomd/hpmc/pytest/test_quick_compress.py:
```diff
@@ -35,8 +50,17 @@
     ('trigger', hoomd.trigger.Before(12345)),
-    ('target_box', hoomd.Box.from_box([10, 20, 30])),
-    ('target_box', hoomd.Box.from_box([50, 50])),
+    ('target_box', hoomd.variant.box.Constant(hoomd.Box.from_box([10, 20,
```
Should I be able to set the target_box attribute to a regular box too?

joaander replied:
Yes, you need to be able to set a regular box both for backwards compatibility and convenience.

tcmoore3 replied:
Agreed with Josh.

tommy-waltmann replied:
A test case still needs to be added for this.

tcmoore3 replied:
You're right. Added.

tommy-waltmann reviewed (changes requested):
This looks pretty good, I just have some small changes to consider.

joaander commented on hoomd/BoxDim.h:
```diff
@@ -601,10 +601,25 @@ struct
         return L1 == L2 && xy1 == xy2 && xz1 == xz2 && yz1 == yz2;
         }
 
+    HOSTDEVICE bool operator==(const std::array<Scalar, 6>& other) const
+        {
+        Scalar3 L = getL();
+        Scalar xy = getTiltFactorXY();
+        Scalar xz = getTiltFactorXZ();
+        Scalar yz = getTiltFactorYZ();
+        std::array<Scalar, 6> box = {L.x, L.y, L.z, xy, xz, yz};
+        return box == other;
+        }
+
```
Let's not introduce unneeded == operators.
```suggestion
```

joaander commented on hoomd/BoxDim.h:
```diff
@@ -601,10 +601,25 @@ struct
         {
         return !((*this) == other);
         }
+
+    HOSTDEVICE bool operator!=(const std::array<Scalar, 6>& other) const
+        {
+        return !((*this) == other);
+        }
```
```suggestion
```

joaander commented on hoomd/hpmc/UpdaterQuickCompress.cc:
```diff
@@ -48,13 +48,13 @@ void UpdaterQuickCompress::update(uint64_t timestep)
     BoxDim current_box = m_pdata->getGlobalBox();
 
-    if (n_overlaps == 0 && current_box != *m_target_box)
+    if (n_overlaps == 0 && current_box != (*m_target_box)(timestep))
```
```suggestion
    BoxDim target_box = BoxDim((*m_target_box)(timestep));
    if (n_overlaps == 0 && current_box != target_box)
```
Use the existing `BoxDim` `!=` check.

I will add a `BoxDim` constructor so this will work.

joaander commented on hoomd/hpmc/UpdaterQuickCompress.cc:
```diff
@@ -48,13 +48,13 @@ void UpdaterQuickCompress::update(uint64_t timestep)
-    if (n_overlaps == 0 && current_box != *m_target_box)
+    if (n_overlaps == 0 && current_box != (*m_target_box)(timestep))
         {
         performBoxScale(timestep);
```
```suggestion
        performBoxScale(timestep, target_box);
```
Add `const BoxDim& target_box` to the arguments of `performBoxScale`.

joaander commented on hoomd/hpmc/UpdaterQuickCompress.cc:
```diff
@@ -48,13 +48,13 @@ void UpdaterQuickCompress::update(uint64_t timestep)
 
     // The compression is complete when we have reached the target box and there are no overlaps.
-    if (n_overlaps == 0 && current_box == *m_target_box)
+    if (n_overlaps == 0 && current_box == (*m_target_box)(timestep))
```
```suggestion
    if (n_overlaps == 0 && current_box == target_box)
```

joaander commented on hoomd/hpmc/UpdaterQuickCompress.cc:
```diff
@@ -183,31 +183,37 @@ BoxDim UpdaterQuickCompress::getNewBox(uint64_t timestep)
     double scale = uniform(rng);
 
-    const auto& target_box = *m_target_box;
+    const auto& target_box_dims = (*m_target_box)(timestep);
```
```suggestion
```
Add `const BoxDim& target_box` to the arguments of `getNewBox`. Also, undo all the changes to `getNewBox` below that are no longer necessary as `target_box` is now set.

joaander commented on hoomd/hpmc/update.py:
```diff
@@ -736,7 +736,8 @@ class QuickCompress(Updater):
             triggered time steps.
 
-        target_box (hoomd.box.box_like): Dimensions of the target box.
+        target_box (hoomd.box.box_like or hoomd.variant.box.BoxVariant):
```
```suggestion
        target_box (`hoomd.box.box_like` or `hoomd.variant.box.BoxVariant`):
```
I think this is necessary for the links to work correctly.

joaander commented on hoomd/hpmc/pytest/test_quick_compress.py:
```diff
@@ -235,6 +259,44 @@ def test_disk_compression(phi, simulation_factory, lattice_snapshot_factory):
+        target_density = scale(sim.timestep) / final_volume + (
+            1 - scale(sim.timestep)) / initial_volume
+        target_volume = 1 / target_density
+        assert sim.state.box.volume >= target_volume
```
Does this test fail if you set `target_box` to the final box?

tcmoore3 replied:
No. It would if it was `sim.state.box.volume > target_volume` but the `>=` handles the case where `target_box` is the final box. I updated the test to check that case.

joaander replied:
Let me clarify: `t_ramp=100` is pretty short.
If you set the target box to the final, this test might still pass (where target_volume) is still determined from the ramp. We need `t_ramp` set large enough so that this test is validly checking that the target box variant is applied.

joaander commented on hoomd/hpmc/UpdaterQuickCompress.cc:
```diff
@@ -183,31 +183,37 @@ BoxDim UpdaterQuickCompress::getNewBox(uint64_t timestep)
 
-    const auto& target_box = *m_target_box;
+    const auto& target_box_dims = (*m_target_box)(timestep);
+    Scalar target_Lx = target_box_dims[0];
+    Scalar target_Ly = target_box_dims[1];
+    Scalar target_Lz = target_box_dims[2];
+    Scalar target_xy = target_box_dims[3];
+    Scalar target_xz = target_box_dims[4];
+    Scalar target_yz = target_box_dims[5];
```
```suggestion
```

joaander reviewed (changes requested):
Thanks! I have many suggestions to reduce the number of changes needed and make this a clean change set.

tcmoore3 committed:
Add box variant preprocess

joaander committed:
Resolve ciruclar import.

tcmoore3 committed:
Apply suggestions from code review

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

joaander committed:
Add BoxDim array constructor.

joaander committed:
Misc fixes.

tcmoore3 committed:
Use target_box for getNewBox and performBoxScale

tcmoore3 committed:
Include header <array> in BoxDim.h

tcmoore3 committed:
Update tests

tcmoore3 committed:
Update slow compress test

tcmoore3 committed:
Remove unused variables

tcmoore3 committed:
formatting

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/quick-compress-moving-target-box

tcmoore3 committed:
Update test

tcmoore3 committed:
Fix precision issues in test

tcmoore3 committed:
format

tommy-waltmann reviewed (changes requested):
I think there is just one more comment of mine from last time to address.

tcmoore3 commented:
I only realized this while addressing [this comment](https://github.com/glotzerlab/hoomd-blue/pull/1736#discussion_r1514948806), but the notion of "complete" for a quick compress updater is somewhat ambiguous now that you can pass in a moving target box. The C++ updater sets `m_is_complete` if the box is equal to the target box (evaluated at the current timestep) and there are no overlaps. This means `quick_compress.complete` returns `True` if the box can easily compress to the setpoint for the specific step without overlaps. This situation occured when I was updating that test. The test contained:
```
# qc.t_ramp = 10000
while not qc.complete:
        sim.run(10)
        simulation_density = 1 / sim.state.box.volume
        target_density = 1 / hoomd.Box(*qc.target_box(sim.timestep)).volume
        assert simulation_density < target_density or (
            simulation_density == pytest.approx(target_density))
assert sim.timestep >= qc.target_box.t_ramp
```
After 10 steps, `qc.complete` returned `True` since the box reached the target for that timestep without overlaps, despite the step being far lower than the ramp length. 

Should we do anything to handle this scenario? It's easy enough to compute final target box for the `BoxVariant`s we implemented so far, but not all variants necessarily have a final box, e.g., I can imagine a `variant.box.Cycle`.  I imagine the typical use case is currently something along the lines of:
```
while not qc.complete:
    sim.run(N)
```
So do we now just let the user handle this, e.g., changing the above to something like
```
while sim.timestep < t_ramp and (not qc.complete):
    sim.run(N)
```
?

tcmoore3 committed:
Update slow compression test

tcmoore3 committed:
Update tests

joaander commented:
> So do we now just let the user handle this, e.g., changing the above to something like
> 
> ```
> while sim.timestep < t_ramp and (not qc.complete):
>     sim.run(N)
> ```

Yes, we should make sure the behavior is documented clearly. I don't want to require that all box variants (including custom ones) provide a valid `complete` attribute. Users should instead adjust their `while` conditions appropriately for the variant of their choice.

tcmoore3 committed:
Add note in docstring about complete

joaander commented on hoomd/hpmc/update.py:
This comment is not related to this pull request, but I noticed the docs are out of date. @janbridley can you please make a small PR to update the docs below. I think the `.. math::` block no longer accurately describes how the box is scaled after your fixes.

janbridley replied:
Good catch! I just pushed `ac17bf8dd` which fixes this. The block is a little harder to read due to the more complicated formulas, but should still do the trick.

joaander commented on hoomd/hpmc/update.py:
```diff
@@ -845,11 +848,25 @@ class QuickCompress(Updater):
+            sim.state.box, sim.state.box.volume / 2, 0, 10000)
+        qc = hoomd.hpmc.update.QuickCompress(10, target_box)
+        while (not qc.complete) and sim.timestep < qc.target_box.t_ramp:
+            sim.run(100)
```
```suggestion
        while sim.timestep < target_box.t_ramp + target_box.t_start or (not qc.complete):
            sim.run(100)
```
Use `or` to continue running past the end of `t_start + t_ramp` until the compression is complete. Also change the order of the lazy evaluation to make it more obvious this while loop is until the end of the ramp.

joaander commented on hoomd/hpmc/update.py:
```diff
@@ -845,11 +848,25 @@ class QuickCompress(Updater):
+    a condition that checks both the `complete` attribute of the updater and the
+    simulation timestep. For example::
+
+        target_box = hoomd.variant.box.InverseVolumeRamp(
+            sim.state.box, sim.state.box.volume / 2, 0, 10000)
```
```suggestion
        target_box = hoomd.variant.box.InverseVolumeRamp(
            sim.state.box, sim.state.box.volume / 2, 0, 1_000)
```
We should put this in Sybil, but I see that there are no other Sybil examples in this file. I'll make a separate PR for a large batch of Sybil updates.

For now, let's shorten the steps so the doc example will not run for long when it is added to Sybil.

joaander reviewed (approved):
Thanks! I have two suggestions to improve the docs. After those changes, this looks great.

joaander commented:
One more fix is needed. `test_inverse_volume_slow_compress[CPU-2]` uses too small of a box for domain decomposition in MPI: https://github.com/glotzerlab/hoomd-blue/actions/runs/8242679743/job/22542233097?pr=1736#step:8:2266

joaander committed:
Apply suggestions from code review

janbridley committed:
Fix `.. math::` block in update.py

tcmoore3 committed:
Update failing test and mark with pytest.validate

`test_inverse_volume_slow_compress` was failing on MPI because the box
was too small for domain decomposition, so this commit uses a larger
system.

That test also takes ~4 seconds to run in serial, so mark it with
pytest.validate since we do not gain anything from running that test
on every commit.

tcmoore3 committed:
Fix formatting in code example

tcmoore3 committed:
Do not include array in BoxDim in JIT builds

tcmoore3 commented:
Resolves #1641
